File: week_1/day-2-prefix-sum-hashing/variations_que/longest_subarray_sum_k.py
# longest_subarray_sum_k.py

def longestSubarray(nums, k):
    prefix = 0
    index_map = {0: -1}
    max_len = 0

    for i, num in enumerate(nums):
        prefix += num
        if prefix - k in index_map:
            max_len = max(max_len, i - index_map[prefix - k])
        if prefix not in index_map:
            index_map[prefix] = i

    return max_len

# Example Input
nums = [1, -1, 5, -2, 3]
k = 3

# Function Call
print(longestSubarray(nums, k))  # Output: 4 (subarray [1, -1, 5, -2])

File: week_1/day-2-prefix-sum-hashing/variations_que/longest_subarray_sum_k.js
// longest_subarray_sum_k.js

function longestSubarray(nums, k) {
    let prefix = 0;
    const map = new Map();
    map.set(0, -1);
    let maxLen = 0;

    for (let i = 0; i < nums.length; i++) {
        prefix += nums[i];
        if (map.has(prefix - k)) {
            maxLen = Math.max(maxLen, i - map.get(prefix - k));
        }
        if (!map.has(prefix)) {
            map.set(prefix, i);
        }
    }
    return maxLen;
}

// Example Input
const nums = [1, -1, 5, -2, 3];
const k = 3;

console.log(longestSubarray(nums, k)); // Output: 4 (subarray [1, -1, 5, -2])
